File: src/main/java/net/easecation/ghosty/recording/entity/updated/EntityUpdatedSkinId.java
package net.easecation.ghosty.recording.entity.updated;

import cn.nukkit.entity.Entity;
import cn.nukkit.entity.data.IntEntityData;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.entity.SimulatedEntity;
import net.easecation.ghosty.recording.entity.EntityRecordNode;

/**
 * Created by Mulan Lin('Snake1999') on 2016/11/19 15:26.
 */
public class EntityUpdatedSkinId implements EntityUpdated {

    private int data;

    @Override
    public void processTo(SimulatedEntity entity) {
        entity.setDataProperty(new IntEntityData(Entity.DATA_SKIN_ID, this.data));
    }

    @Override
    public EntityRecordNode applyTo(EntityRecordNode node) {
        node.setSkinId(data);
        return node;
    }

    public static EntityUpdatedSkinId of(int data) {
        return new EntityUpdatedSkinId(data);
    }

    @Override
    public int getUpdateTypeId() {
        return TYPE_SKIN_ID;
    }

    @Override
    public boolean hasStates() {
        return true;
    }

    public EntityUpdatedSkinId(BinaryStream stream) {
        this.read(stream);
    }

    private EntityUpdatedSkinId(int data) {
        this.data = data;
    }

    @Override
    public boolean equals(Object obj) {
        if(!(obj instanceof EntityUpdatedSkinId o)) return false;
        return data == o.data;
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putVarInt(this.data);
    }

    @Override
    public void read(BinaryStream stream) {
        this.data = stream.getVarInt();
    }

    @Override
    public String toString() {
        return "EntityUpdatedSkinId{" +
            "data=" + data +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/entity/updated/EntityUpdatedSkinId.kt
package net.easecation.ghosty.recording.entity.updated

import cn.nukkit.entity.Entity
import cn.nukkit.entity.data.IntEntityData
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.entity.SimulatedEntity
import net.easecation.ghosty.recording.UpdateWithState
import net.easecation.ghosty.recording.entity.EntityRecordNode

@Serializable
data class EntityUpdatedSkinId(
    val data: Int
) : EntityUpdated, UpdateWithState {
    override fun getUpdateTypeId(): Int = EntityUpdated.TYPE_SKIN_ID
    override fun processTo(entity: SimulatedEntity) {
        entity.setDataProperty(IntEntityData(Entity.DATA_SKIN_ID, data))
    }

    override fun applyTo(node: EntityRecordNode): EntityRecordNode {
        node.skinId = data
        return node
    }

    companion object {
        @JvmStatic
        fun of(data: Int) = EntityUpdatedSkinId(data)

        @JvmStatic
        val ADAPTER: UpdateAdapter<EntityUpdatedSkinId> = Adapter
    }

    private object Adapter : UpdateAdapter<EntityUpdatedSkinId> {
        override fun hasStates() = true
        override fun write(updated: EntityUpdatedSkinId, stream: BinaryStream) {
            stream.putVarInt(updated.data)
        }
        override fun read(stream: BinaryStream) = EntityUpdatedSkinId(
            data = stream.varInt
        )
    }
}
